mdf_matio/validator.py

```diff
@@ -0,0 +1,397 @@
+                }
+
+        # Create temporary file for records
+        self.__tempfile = TemporaryFile(mode="w+")
+
+        # Save dataset metadata
+        # Also ensure metadata is JSON-serializable
+        self.__dataset = json.loads(json.dumps(ds_md))
+
+        # Return results
+        return {
+            "success": True
+            }
+
+    def add_record(self, rc_md):
+        """Validate a record against the MDF schema.
+
+        Arguments:
+        rc_md (dict): The record metadata to validate.
+
+        Returns:
+        dict: success (bool): True on success, False on failure
+            If success is False:
+              error (str): A short message about the error.
+              details (str): The full jsonschema error message.
+        """
+        if self.__finished:
+            return {
+                "success": False,
+                "error": ("Dataset has been finished by calling get_finished_dataset(),"
+                          " and no more records may be entered.")
+                }
+        elif not self.__dataset:
+            return {
+                "success": False,
+                "error": "Dataset not started."
+                }
+
+        # Load schema
+        with open(os.path.join(self.__schema_dir, "record.json")) as schema_file:
+            schema = json.load(schema_file)
+        resolver = jsonschema.RefResolver(base_uri="file://{}/".format(self.__schema_dir),
+                                          referrer=schema)
+
+        # Add any missing blocks
+        if not rc_md.get("mdf"):
+            rc_md["mdf"] = {}
+        if not rc_md.get("files"):
+            rc_md["files"] = []
+        elif isinstance(rc_md["files"], dict):
+            rc_md["files"] = [rc_md["files"]]
+        if not rc_md.get("material"):
+            rc_md["material"] = {}
+
+        # Add fields
+        # BLOCK: mdf
+        # source_id
+        rc_md["mdf"]["source_id"] = self.__dataset["mdf"]["source_id"]
+
+        # source_name
+        rc_md["mdf"]["source_name"] = self.__dataset["mdf"]["source_name"]
+
+        # scroll_id
+        rc_md["mdf"]["scroll_id"] = self.__scroll_id
+        self.__scroll_id += 1
+
+        # ingest_date
+        rc_md["mdf"]["ingest_date"] = self.__ingest_date
+
+        # resource_type
+        rc_md["mdf"]["resource_type"] = "record"
+
+        # version
+        rc_md["mdf"]["version"] = self.__dataset["mdf"]["version"]
+
+        # acl
+        if not rc_md["mdf"].get("acl"):
+            rc_md["mdf"]["acl"] = self.__base_acl or self.__dataset["mdf"]["acl"]
+
+        # organizations
+        if self.__dataset["mdf"].get("organizations"):
+            rc_md["mdf"]["organizations"] = self.__dataset["mdf"]["organizations"]
+
+        # BLOCK: files
+        # Add file data to dataset
+        if rc_md["files"]:
+            self.__indexed_files += rc_md["files"]
+
+        # BLOCK: material
+        # elements
+        if rc_md["material"].get("composition"):
```

Comment:
This code seems like a candidate for breaking out into a separate function.

Reply:
Why is that? It's not reused anywhere, and is (reasonably) simple. I'm aware there are several commented-out blocks, that have been previous approaches I want to keep around to turn back on if necessary.

Reply:
My reasoning would be to make it easier to test by being able to invoke it in unit tests without engaging all the other parts of this code. That and you could document the desired behavior of the code in a place where users are more likely to find it.

Reply:
It's easy enough to have tests that engage this code (records with `material.composition` specified); I'm not convinced there's a significant testing benefit. Additionally, users should not be using this code directly anyway.